// rsuite/lib/Animation/propTypes.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var propTypes = require('../../../../_virtual/propTypes2.js_commonjs-exports');
require('../../../@babel/runtime/helpers/interopRequireDefault.js');
require('../../../prop-types/index.js');
var index = require('../../../../_virtual/index3.js_commonjs-module');
var interopRequireDefault = require('../../../../_virtual/interopRequireDefault.js_commonjs-module');

var _interopRequireDefault = interopRequireDefault.interopRequireDefault.exports;

propTypes.__exports.__esModule = true;
propTypes.__exports.animationPropTypes = void 0;

var _propTypes = _interopRequireDefault(index.propTypes.exports);

var animationPropTypes = {
  onEnter: _propTypes.default.func,
  onEntering: _propTypes.default.func,
  onEntered: _propTypes.default.func,
  onExit: _propTypes.default.func,
  onExiting: _propTypes.default.func,
  onExited: _propTypes.default.func
};
propTypes.__exports.animationPropTypes = animationPropTypes;

exports.default = propTypes.__exports;
