// dom-lib/es/animation/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var canUseDOM = require('../query/canUseDOM.js');
var cancelAnimationFramePolyfill = require('./cancelAnimationFramePolyfill.js');
var nativeRequestAnimationFrame = require('./nativeRequestAnimationFrame.js');
var requestAnimationFramePolyfill = require('./requestAnimationFramePolyfill.js');

var vendorMap = {
  animation: 'animationend',
  OAnimation: 'oAnimationEnd',
  MozAnimation: 'animationend',
  WebkitAnimation: 'webkitAnimationEnd'
};

function getAnimationEvent() {
  if (!canUseDOM['default']) {
    return;
  }

  var tempAnimationEnd;
  var style = document.createElement('div').style;

  for (tempAnimationEnd in vendorMap) {
    if (style[tempAnimationEnd] !== undefined) {
      return vendorMap[tempAnimationEnd];
    }
  }
}

var events = function events() {
  return {
    end: getAnimationEvent()
  };
};

exports.cancelAnimationFramePolyfill = cancelAnimationFramePolyfill['default'];
exports.nativeRequestAnimationFrame = nativeRequestAnimationFrame['default'];
exports.requestAnimationFramePolyfill = requestAnimationFramePolyfill['default'];
exports.events = events;
