// rsuite-table/es/CellGroup.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('../../@babel/runtime/helpers/esm/extends.js');
var objectWithoutPropertiesLoose = require('../../@babel/runtime/helpers/esm/objectWithoutPropertiesLoose.js');
var inheritsLoose = require('../../@babel/runtime/helpers/esm/inheritsLoose.js');
var React = require('react');
var index = require('../../prop-types/index.js');
var cx = require('classnames');
var prefix = require('./utils/prefix.js');
var getUnhandledProps = require('./utils/getUnhandledProps.js');
require('../../lodash/isPlainObject.js');
require('../../lodash/_baseGetTag.js');
require('../../lodash/_stringToPath.js');
require('../../lodash/_baseToString.js');
require('./ColumnGroup.js');
require('./HeaderCell.js');
require('../../dom-lib/es/animation/cancelAnimationFramePolyfill.js');
require('../../dom-lib/es/animation/nativeRequestAnimationFrame.js');
require('../../dom-lib/es/animation/requestAnimationFramePolyfill.js');
require('../../dom-lib/es/transition/translateDOMPositionXY.js');
require('../../dom-lib/es/getVendorPrefixedName.js');
require('../../dom-lib/es/utils/emptyFunction.js');
require('../../dom-lib/es/normalizeWheel.js');
require('./utils/translateDOMPositionXY.js');
require('../../lodash/intersection.js');
require('../../dom-lib/lib/animation/requestAnimationFramePolyfill.js');
require('../../dom-lib/lib/animation/cancelAnimationFramePolyfill.js');
var TableContext = require('./TableContext.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

var propTypes = {
  fixed: index['default'].oneOf(['left', 'right']),
  width: index['default'].number,
  height: index['default'].number,
  left: index['default'].number,
  style: index['default'].object,
  className: index['default'].string,
  classPrefix: index['default'].string
};

var CellGroup = /*#__PURE__*/function (_React$PureComponent) {
  inheritsLoose['default'](CellGroup, _React$PureComponent);

  function CellGroup() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$PureComponent.call.apply(_React$PureComponent, [this].concat(args)) || this;

    _this.addPrefix = function (name) {
      return prefix.prefix(_this.props.classPrefix)(name);
    };

    return _this;
  }

  var _proto = CellGroup.prototype;

  _proto.render = function render() {
    var _classNames;

    var _this$props = this.props,
        fixed = _this$props.fixed,
        width = _this$props.width,
        left = _this$props.left,
        height = _this$props.height,
        style = _this$props.style,
        classPrefix = _this$props.classPrefix,
        className = _this$props.className,
        children = _this$props.children,
        rest = objectWithoutPropertiesLoose['default'](_this$props, ["fixed", "width", "left", "height", "style", "classPrefix", "className", "children"]);

    var classes = cx__default['default'](classPrefix, className, (_classNames = {}, _classNames[this.addPrefix("fixed-" + (fixed || ''))] = fixed, _classNames[this.addPrefix('scroll')] = !fixed, _classNames));

    var styles = _extends['default']({
      width: width,
      height: height
    }, style);

    var unhandledProps = getUnhandledProps['default'](propTypes, rest);
    return /*#__PURE__*/React__namespace.createElement(TableContext['default'].Consumer, null, function (_ref) {
      var translateDOMPositionXY = _ref.translateDOMPositionXY;
      translateDOMPositionXY === null || translateDOMPositionXY === void 0 ? void 0 : translateDOMPositionXY(styles, left, 0);
      return /*#__PURE__*/React__namespace.createElement("div", _extends['default']({}, unhandledProps, {
        className: classes,
        style: styles
      }), children);
    });
  };

  return CellGroup;
}(React__namespace.PureComponent);

CellGroup.propTypes = propTypes;
CellGroup.defaultProps = {
  classPrefix: prefix.defaultClassPrefix('table-cell-group')
};

exports['default'] = CellGroup;
