// rsuite-table/es/ColumnResizeHandler.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('../../@babel/runtime/helpers/esm/extends.js');
var objectWithoutPropertiesLoose = require('../../@babel/runtime/helpers/esm/objectWithoutPropertiesLoose.js');
var inheritsLoose = require('../../@babel/runtime/helpers/esm/inheritsLoose.js');
var React = require('react');
var index = require('../../prop-types/index.js');
var cx = require('classnames');
var clamp = require('../../lodash/clamp.js');
require('../../dom-lib/es/animation/cancelAnimationFramePolyfill.js');
require('../../dom-lib/es/animation/nativeRequestAnimationFrame.js');
require('../../dom-lib/es/animation/requestAnimationFramePolyfill.js');
require('../../dom-lib/es/transition/translateDOMPositionXY.js');
require('../../dom-lib/es/getVendorPrefixedName.js');
var DOMMouseMoveTracker = require('../../dom-lib/es/DOMMouseMoveTracker.js');
require('../../dom-lib/es/utils/emptyFunction.js');
require('../../dom-lib/es/normalizeWheel.js');
var prefix = require('./utils/prefix.js');
var getUnhandledProps = require('./utils/getUnhandledProps.js');
require('../../lodash/isPlainObject.js');
require('../../lodash/_baseGetTag.js');
require('../../lodash/_stringToPath.js');
require('../../lodash/_baseToString.js');
require('./ColumnGroup.js');
require('./HeaderCell.js');
require('./utils/translateDOMPositionXY.js');
require('../../lodash/intersection.js');
require('../../dom-lib/lib/animation/requestAnimationFramePolyfill.js');
require('../../dom-lib/lib/animation/cancelAnimationFramePolyfill.js');
var TableContext = require('./TableContext.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

var propTypes = {
  height: index['default'].number,
  defaultColumnWidth: index['default'].number,
  columnLeft: index['default'].number,
  columnFixed: index['default'].oneOfType([index['default'].bool, index['default'].oneOf(['left', 'right'])]),
  className: index['default'].string,
  classPrefix: index['default'].string,
  style: index['default'].object,
  onColumnResizeStart: index['default'].func,
  onColumnResizeEnd: index['default'].func,
  onColumnResizeMove: index['default'].func
};

var ColumnResizeHandler = /*#__PURE__*/function (_React$Component) {
  inheritsLoose['default'](ColumnResizeHandler, _React$Component);

  function ColumnResizeHandler(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.columnWidth = 0;
    _this.cursorDelta = 0;
    _this.mouseMoveTracker = void 0;
    _this.isKeyDown = void 0;

    _this.onMove = function (deltaX) {
      if (!_this.isKeyDown) {
        return;
      }

      var _this$props = _this.props,
          onColumnResizeMove = _this$props.onColumnResizeMove,
          defaultColumnWidth = _this$props.defaultColumnWidth,
          columnLeft = _this$props.columnLeft,
          columnFixed = _this$props.columnFixed;
      var rtl = _this.context.rtl;
      _this.cursorDelta += deltaX;
      _this.columnWidth = clamp['default'](defaultColumnWidth + (rtl ? -_this.cursorDelta : _this.cursorDelta), 20, 20000);
      onColumnResizeMove === null || onColumnResizeMove === void 0 ? void 0 : onColumnResizeMove(_this.columnWidth, columnLeft, columnFixed);
    };

    _this.onColumnResizeEnd = function () {
      var _this$props$onColumnR, _this$props2, _this$mouseMoveTracke, _this$mouseMoveTracke2;

      _this.isKeyDown = false;
      (_this$props$onColumnR = (_this$props2 = _this.props).onColumnResizeEnd) === null || _this$props$onColumnR === void 0 ? void 0 : _this$props$onColumnR.call(_this$props2, _this.columnWidth, _this.cursorDelta);
      (_this$mouseMoveTracke = _this.mouseMoveTracker) === null || _this$mouseMoveTracke === void 0 ? void 0 : (_this$mouseMoveTracke2 = _this$mouseMoveTracke.releaseMouseMoves) === null || _this$mouseMoveTracke2 === void 0 ? void 0 : _this$mouseMoveTracke2.call(_this$mouseMoveTracke);
      _this.mouseMoveTracker = null;
    };

    _this.onColumnResizeMouseDown = function (event) {
      var _this$props$onColumnR2, _this$props3;

      _this.mouseMoveTracker = _this.getMouseMoveTracker();

      _this.mouseMoveTracker.captureMouseMoves(event);

      _this.isKeyDown = true;
      _this.cursorDelta = 0;
      var client = {
        clientX: event.clientX,
        clientY: event.clientY,
        preventDefault: Function()
      };
      (_this$props$onColumnR2 = (_this$props3 = _this.props).onColumnResizeStart) === null || _this$props$onColumnR2 === void 0 ? void 0 : _this$props$onColumnR2.call(_this$props3, client);
    };

    _this.columnWidth = props.defaultColumnWidth || 0;
    return _this;
  }

  var _proto = ColumnResizeHandler.prototype;

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.mouseMoveTracker) {
      this.mouseMoveTracker.releaseMouseMoves();
      this.mouseMoveTracker = null;
    }
  };

  _proto.getMouseMoveTracker = function getMouseMoveTracker() {
    return this.mouseMoveTracker || new DOMMouseMoveTracker['default'](this.onMove, this.onColumnResizeEnd, document.body);
  };

  _proto.render = function render() {
    var _extends2;

    var _this$props4 = this.props,
        _this$props4$columnLe = _this$props4.columnLeft,
        columnLeft = _this$props4$columnLe === void 0 ? 0 : _this$props4$columnLe,
        classPrefix = _this$props4.classPrefix,
        height = _this$props4.height,
        className = _this$props4.className,
        style = _this$props4.style,
        columnFixed = _this$props4.columnFixed,
        rest = objectWithoutPropertiesLoose['default'](_this$props4, ["columnLeft", "classPrefix", "height", "className", "style", "columnFixed"]);

    if (columnFixed === 'right') {
      return null;
    }

    var rtl = this.context.rtl;

    var styles = _extends['default']((_extends2 = {}, _extends2[rtl ? 'right' : 'left'] = this.columnWidth + columnLeft - 2, _extends2.height = height, _extends2), style);

    var classes = cx__default['default'](classPrefix, className);
    var unhandled = getUnhandledProps['default'](propTypes, rest);
    return /*#__PURE__*/React__namespace.createElement("div", _extends['default']({}, unhandled, {
      className: classes,
      style: styles,
      onMouseDown: this.onColumnResizeMouseDown,
      role: "button",
      tabIndex: -1
    }));
  };

  return ColumnResizeHandler;
}(React__namespace.Component);

ColumnResizeHandler.contextType = TableContext['default'];
ColumnResizeHandler.propTypes = propTypes;
ColumnResizeHandler.defaultProps = {
  classPrefix: prefix.defaultClassPrefix('table-column-resize-spanner')
};

exports['default'] = ColumnResizeHandler;
