// rsuite-table/es/ColumnGroup.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('../../@babel/runtime/helpers/esm/extends.js');
var objectWithoutPropertiesLoose = require('../../@babel/runtime/helpers/esm/objectWithoutPropertiesLoose.js');
var React = require('react');
var cx = require('classnames');
var index = require('../../prop-types/index.js');
var prefix = require('./utils/prefix.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

var ColumnGroup = /*#__PURE__*/React__namespace.forwardRef(function (props, ref) {
  var header = props.header,
      className = props.className,
      children = props.children,
      classPrefix = props.classPrefix,
      headerHeight = props.headerHeight,
      verticalAlign = props.verticalAlign,
      width = props.width,
      rest = objectWithoutPropertiesLoose['default'](props, ["header", "className", "children", "classPrefix", "headerHeight", "verticalAlign", "width"]);

  var height = headerHeight / 2;
  var styles = {
    height: height,
    width: width
  };

  var contentStyles = _extends['default']({}, styles, {
    verticalAlign: verticalAlign
  });

  var addPrefix = React__namespace.useCallback(function (name) {
    return prefix.prefix(classPrefix)(name);
  }, []);
  return /*#__PURE__*/React__namespace.createElement("div", _extends['default']({
    ref: ref,
    className: cx__default['default'](classPrefix, className)
  }, rest), /*#__PURE__*/React__namespace.createElement("div", {
    className: addPrefix('header'),
    style: styles
  }, /*#__PURE__*/React__namespace.createElement("div", {
    className: addPrefix('header-content'),
    style: contentStyles
  }, header)), React__namespace.Children.map(children, function (node) {
    var _node$props;

    var nodeStyles = _extends['default']({
      height: height
    }, (_node$props = node.props) === null || _node$props === void 0 ? void 0 : _node$props.style, {
      top: styles.height
    });

    return /*#__PURE__*/React__namespace.cloneElement(node, {
      className: addPrefix('cell'),
      style: nodeStyles,
      headerHeight: height,
      verticalAlign: verticalAlign,
      children: /*#__PURE__*/React__namespace.createElement("span", {
        className: addPrefix('cell-content')
      }, node.props.children)
    });
  }));
});
ColumnGroup.displayName = 'ColumnGroup';
ColumnGroup.defaultProps = {
  headerHeight: 80,
  classPrefix: prefix.defaultClassPrefix('table-column-group')
};
ColumnGroup.propTypes = {
  header: index['default'].node,
  classPrefix: index['default'].string,
  verticalAlign: index['default'].oneOf(['top', 'middle', 'bottom'])
};

exports['default'] = ColumnGroup;
