// rsuite-table/es/ColumnGroup.js
import _extends from '../../@babel/runtime/helpers/esm/extends.js';
import _objectWithoutPropertiesLoose from '../../@babel/runtime/helpers/esm/objectWithoutPropertiesLoose.js';
import React__default from 'react';
import PropTypes from '../../prop-types/index.js';
import './utils/prefix.js';
import '../../lodash/isPlainObject.js';
import '../../lodash/_baseGetTag.js';
import '../../lodash/_stringToPath.js';
import '../../lodash/_baseToString.js';
import './HeaderCell.js';
import './utils/translateDOMPositionXY.js';
import '../../dom-lib/esm/requestAnimationFramePolyfill.js';
import '../../dom-lib/esm/cancelAnimationFramePolyfill.js';
import useClassNames from './utils/useClassNames.js';
import '../../react-is/index.js';
import '../../lodash/_isFlattenable.js';
import '../../lodash/omit.js';
import '../../lodash/merge.js';
import '../../@juggle/resize-observer/lib/ResizeObserver.js';
import '../../@juggle/resize-observer/lib/algorithms/calculateBoxSize.js';
import '../../lodash/_getTag.js';
import '../../lodash/isArguments.js';
import '../../lodash/isBuffer.js';
import '../../lodash/isTypedArray.js';
import '../../dom-lib/esm/utils/emptyFunction.js';
import '../../dom-lib/esm/utils/normalizeWheel.js';

var ColumnGroup = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var header = props.header,
      className = props.className,
      children = props.children,
      classPrefix = props.classPrefix,
      headerHeight = props.headerHeight,
      verticalAlign = props.verticalAlign,
      align = props.align,
      width = props.width,
      groupHeightProp = props.groupHeaderHeight,
      rest = _objectWithoutPropertiesLoose(props, ["header", "className", "children", "classPrefix", "headerHeight", "verticalAlign", "align", "width", "groupHeaderHeight"]);

  var hasGroupHeight = typeof groupHeightProp !== 'undefined';
  var groupHeight = hasGroupHeight ? groupHeightProp : headerHeight / 2;
  var restHeight = hasGroupHeight ? headerHeight - groupHeightProp : headerHeight / 2;
  var styles = {
    height: groupHeight,
    width: width
  };

  var _useClassNames = useClassNames(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix,
      merge = _useClassNames.merge,
      prefix = _useClassNames.prefix;

  var classes = merge(className, withClassPrefix());

  var contentStyles = _extends({}, styles, {
    textAlign: align,
    verticalAlign: verticalAlign
  });

  return /*#__PURE__*/React__default.createElement("div", _extends({
    ref: ref,
    className: classes
  }, rest), /*#__PURE__*/React__default.createElement("div", {
    className: prefix('header'),
    style: styles
  }, /*#__PURE__*/React__default.createElement("div", {
    className: prefix('header-content'),
    style: contentStyles
  }, header)), React__default.Children.map(children, function (node) {
    return /*#__PURE__*/React__default.cloneElement(node, {
      className: prefix('cell'),
      predefinedStyle: {
        height: restHeight,
        top: styles.height
      },
      headerHeight: restHeight,
      verticalAlign: node.props.verticalAlign || verticalAlign,
      children: /*#__PURE__*/React__default.createElement("span", {
        className: prefix('cell-content')
      }, node.props.children)
    });
  }));
});
ColumnGroup.displayName = 'Table.ColumnGroup';
ColumnGroup.defaultProps = {
  headerHeight: 80,
  classPrefix: 'column-group'
};
ColumnGroup.propTypes = {
  header: PropTypes.node,
  classPrefix: PropTypes.string,
  groupHeaderHeight: PropTypes.number,
  verticalAlign: PropTypes.oneOf(['top', 'middle', 'bottom'])
};

export { ColumnGroup as default };
