// rsuite/lib/constants.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var constants = require('../../../_virtual/constants.js_commonjs-exports');

constants.__exports.__esModule = true;
constants.__exports.DATERANGE_DISABLED_TARGET = constants.__exports.KEY_CODE = constants.__exports.TREE_NODE_DROP_POSITION = constants.__exports.TREE_NODE_ROOT_PADDING = constants.__exports.TREE_NODE_PADDING = constants.__exports.CHECK_STATE = constants.__exports.PLACEMENT = constants.__exports.PLACEMENT_AUTO = constants.__exports.PLACEMENT_8 = constants.__exports.PLACEMENT_4 = constants.__exports.COLOR = constants.__exports.STATUS = constants.__exports.SIZE = constants.__exports.PAGINATION_ICON_NAMES = constants.__exports.STATUS_ICON_NAMES = void 0;
var STATUS_ICON_NAMES = {
  info: 'info',
  success: 'check-circle',
  error: 'close-circle',
  warning: 'remind'
};
constants.__exports.STATUS_ICON_NAMES = STATUS_ICON_NAMES;
var PAGINATION_ICON_NAMES;
constants.__exports.PAGINATION_ICON_NAMES = PAGINATION_ICON_NAMES;

(function (PAGINATION_ICON_NAMES) {
  PAGINATION_ICON_NAMES["more"] = "more";
  PAGINATION_ICON_NAMES["prev"] = "page-previous";
  PAGINATION_ICON_NAMES["next"] = "page-next";
  PAGINATION_ICON_NAMES["first"] = "page-top";
  PAGINATION_ICON_NAMES["last"] = "page-end";
})(PAGINATION_ICON_NAMES || (constants.__exports.PAGINATION_ICON_NAMES = PAGINATION_ICON_NAMES = {}));

var SIZE = ['lg', 'md', 'sm', 'xs'];
constants.__exports.SIZE = SIZE;
var STATUS = ['success', 'warning', 'error', 'info'];
constants.__exports.STATUS = STATUS;
var COLOR = ['red', 'orange', 'yellow', 'green', 'cyan', 'blue', 'violet'];
constants.__exports.COLOR = COLOR;
var PLACEMENT_4 = ['top', 'bottom', 'right', 'left'];
constants.__exports.PLACEMENT_4 = PLACEMENT_4;
var PLACEMENT_8 = ['bottomStart', 'bottomEnd', 'topStart', 'topEnd', 'leftStart', 'rightStart', 'leftEnd', 'rightEnd'];
constants.__exports.PLACEMENT_8 = PLACEMENT_8;
var PLACEMENT_AUTO = ['auto', 'autoVerticalStart', 'autoVerticalEnd', 'autoHorizontalStart', 'autoHorizontalEnd'];
constants.__exports.PLACEMENT_AUTO = PLACEMENT_AUTO;
var PLACEMENT = [].concat(PLACEMENT_4, PLACEMENT_8, PLACEMENT_AUTO);
/**
 *  Check Tree Node State
 */

constants.__exports.PLACEMENT = PLACEMENT;
var CHECK_STATE;
constants.__exports.CHECK_STATE = CHECK_STATE;

(function (CHECK_STATE) {
  CHECK_STATE[CHECK_STATE["UNCHECK"] = 0] = "UNCHECK";
  CHECK_STATE[CHECK_STATE["CHECK"] = 1] = "CHECK";
  CHECK_STATE[CHECK_STATE["INDETERMINATE"] = 2] = "INDETERMINATE";
})(CHECK_STATE || (constants.__exports.CHECK_STATE = CHECK_STATE = {}));

var TREE_NODE_PADDING = 16;
constants.__exports.TREE_NODE_PADDING = TREE_NODE_PADDING;
var TREE_NODE_ROOT_PADDING = 12;
/**
 * Tree Node Drag Type
 */

constants.__exports.TREE_NODE_ROOT_PADDING = TREE_NODE_ROOT_PADDING;
var TREE_NODE_DROP_POSITION;
constants.__exports.TREE_NODE_DROP_POSITION = TREE_NODE_DROP_POSITION;

(function (TREE_NODE_DROP_POSITION) {
  TREE_NODE_DROP_POSITION[TREE_NODE_DROP_POSITION["DRAG_OVER"] = 0] = "DRAG_OVER";
  TREE_NODE_DROP_POSITION[TREE_NODE_DROP_POSITION["DRAG_OVER_TOP"] = 1] = "DRAG_OVER_TOP";
  TREE_NODE_DROP_POSITION[TREE_NODE_DROP_POSITION["DRAG_OVER_BOTTOM"] = 2] = "DRAG_OVER_BOTTOM";
})(TREE_NODE_DROP_POSITION || (constants.__exports.TREE_NODE_DROP_POSITION = TREE_NODE_DROP_POSITION = {}));

var KEY_CODE = {
  BACKSPACE: 8,
  ENTER: 13,
  ESC: 27,
  LEFT: 37,
  UP: 38,
  RIGHT: 39,
  DOWN: 40
};
constants.__exports.KEY_CODE = KEY_CODE;
var DATERANGE_DISABLED_TARGET;
constants.__exports.DATERANGE_DISABLED_TARGET = DATERANGE_DISABLED_TARGET;

(function (DATERANGE_DISABLED_TARGET) {
  DATERANGE_DISABLED_TARGET["CALENDAR"] = "CALENDAR";
  DATERANGE_DISABLED_TARGET["TOOLBAR_BUTTON_OK"] = "TOOLBAR_BUTTON_OK";
  DATERANGE_DISABLED_TARGET["TOOLBAR_SHORTCUT"] = "TOOLBAR_SHORTCUT";
})(DATERANGE_DISABLED_TARGET || (constants.__exports.DATERANGE_DISABLED_TARGET = DATERANGE_DISABLED_TARGET = {}));

exports.default = constants.__exports;
