// rsuite/lib/Picker/MenuWrapper.js
'use strict';

var MenuWrapper = require('../../../../_virtual/MenuWrapper.js_commonjs-module');
require('../../../@babel/runtime/helpers/interopRequireWildcard.js');
require('../../../@babel/runtime/helpers/interopRequireDefault.js');
require('../../../@babel/runtime/helpers/extends.js');
require('../../../@babel/runtime/helpers/objectWithoutPropertiesLoose.js');
require('../../../@babel/runtime/helpers/inheritsLoose.js');
var omit = require('../../../lodash/omit.js');
var React = require('react');
var cx = require('classnames');
require('../../../dom-lib/es/animation/cancelAnimationFramePolyfill.js');
require('../../../dom-lib/es/animation/nativeRequestAnimationFrame.js');
require('../../../dom-lib/es/animation/requestAnimationFramePolyfill.js');
require('../../../dom-lib/es/transition/translateDOMPositionXY.js');
require('../../../dom-lib/es/getVendorPrefixedName.js');
require('../../../dom-lib/es/utils/emptyFunction.js');
require('../../../dom-lib/es/normalizeWheel.js');
var index$1 = require('../../../../_virtual/index2.js_commonjs-proxy');
require('../utils/index.js');
require('../../../element-resize-event/index.js');
require('../utils/getDOMNode.js');
require('../utils/mergeRefs.js');
var index$2 = require('../../../../_virtual/index.js_commonjs-exports');
var _extends = require('../../../../_virtual/extends.js_commonjs-module');
var objectWithoutPropertiesLoose = require('../../../../_virtual/objectWithoutPropertiesLoose.js_commonjs-module');
var inheritsLoose = require('../../../../_virtual/inheritsLoose.js_commonjs-module');
var index = require('../../../../_virtual/index6.js_commonjs-module');
var getDOMNode = require('../../../../_virtual/getDOMNode.js_commonjs-module');
var mergeRefs = require('../../../../_virtual/mergeRefs.js_commonjs-module');
var interopRequireWildcard = require('../../../../_virtual/interopRequireWildcard.js_commonjs-module');
var interopRequireDefault = require('../../../../_virtual/interopRequireDefault.js_commonjs-module');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

(function (module, exports) {

var _interopRequireWildcard = interopRequireWildcard.interopRequireWildcard.exports;

var _interopRequireDefault = interopRequireDefault.interopRequireDefault.exports;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(_extends._extends.exports);

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(objectWithoutPropertiesLoose.objectWithoutPropertiesLoose.exports);

var _inheritsLoose2 = _interopRequireDefault(inheritsLoose.inheritsLoose.exports);

var _omit2 = _interopRequireDefault(omit['default']);

var React = _interopRequireWildcard(React__default['default']);

var _classnames = _interopRequireDefault(cx__default['default']);

var _domLib = index$1['default'];

var _utils = index$2.__exports;

var _elementResizeEvent = _interopRequireWildcard(index.elementResizeEvent.exports);

var _getDOMNode = _interopRequireDefault(getDOMNode.getDOMNode.exports);

var _mergeRefs = _interopRequireDefault(mergeRefs.mergeRefs.exports);

var omitProps = ['placement', 'shouldUpdatePosition', 'arrowOffsetLeft', 'arrowOffsetTop', 'positionLeft', 'positionTop', 'getPositionInstance', 'getToggleInstance', 'autoWidth'];
var resizePlacement = ['topStart', 'topEnd', 'leftEnd', 'rightEnd', 'auto', 'autoVerticalStart', 'autoVerticalEnd', 'autoHorizontalEnd'];

var MenuWrapper =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(MenuWrapper, _React$Component);

  function MenuWrapper(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.menuElementRef = void 0;

    _this.handleResize = function () {
      var getPositionInstance = _this.props.getPositionInstance;
      var instance = getPositionInstance ? getPositionInstance() : null;

      if (instance) {
        instance.updatePosition(true);
      }
    };

    _this.menuElementRef = React.createRef();
    return _this;
  }

  var _proto = MenuWrapper.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var autoWidth = this.props.autoWidth;

    if (resizePlacement.includes(this.props.placement)) {
      (0, _elementResizeEvent.default)(this.menuElementRef.current, this.handleResize);
    }

    if (autoWidth) {
      this.updateMenuStyle();
    }
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.menuElementRef.current) {
      (0, _elementResizeEvent.unbind)(this.menuElementRef.current);
    }
  };

  _proto.updateMenuStyle = function updateMenuStyle() {
    var getToggleInstance = this.props.getToggleInstance;

    if (this.menuElementRef.current && getToggleInstance) {
      var _instance$toggleRef;

      var instance = getToggleInstance();

      if (instance === null || instance === void 0 ? void 0 : (_instance$toggleRef = instance.toggleRef) === null || _instance$toggleRef === void 0 ? void 0 : _instance$toggleRef.current) {
        var width = (0, _domLib.getWidth)((0, _getDOMNode.default)(instance.toggleRef.current));
        (0, _domLib.addStyle)(this.menuElementRef.current, 'min-width', width + "px");
      }
    }
  };

  _proto.render = function render() {
    var _this$props = this.props,
        className = _this$props.className,
        classPrefix = _this$props.classPrefix,
        htmlElementRef = _this$props.htmlElementRef,
        rest = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["className", "classPrefix", "htmlElementRef"]);
    return React.createElement("div", (0, _extends2.default)({}, (0, _omit2.default)(rest, omitProps), {
      ref: (0, _mergeRefs.default)(this.menuElementRef, htmlElementRef),
      className: (0, _classnames.default)(classPrefix, className)
    }));
  };

  return MenuWrapper;
}(React.Component);

var enhance = (0, _utils.defaultProps)({
  classPrefix: 'picker-menu'
});

var _default = enhance(MenuWrapper);

exports.default = _default;
module.exports = exports.default;
}(MenuWrapper.MenuWrapper, MenuWrapper.MenuWrapper.exports));
