// rsuite/lib/Picker/propTypes.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var propTypes = require('../../../../_virtual/propTypes.js_commonjs-exports');
require('../../../@babel/runtime/helpers/interopRequireDefault.js');
require('../../../@babel/runtime/helpers/extends.js');
require('../../../prop-types/index.js');
require('../constants.js');
require('../utils/index.js');
require('../Animation/propTypes.js');
var constants = require('../../../../_virtual/constants.js_commonjs-exports');
var index$1 = require('../../../../_virtual/index.js_commonjs-exports');
var propTypes$1 = require('../../../../_virtual/propTypes2.js_commonjs-exports');
var _extends = require('../../../../_virtual/extends.js_commonjs-module');
var index = require('../../../../_virtual/index5.js_commonjs-module');
var interopRequireDefault = require('../../../../_virtual/interopRequireDefault.js_commonjs-module');

var _interopRequireDefault = interopRequireDefault.interopRequireDefault.exports;

propTypes.__exports.__esModule = true;
propTypes.__exports.listPickerDefaultProps = propTypes.__exports.pickerDefaultProps = propTypes.__exports.listPickerPropTypes = propTypes.__exports.pickerPropTypes = void 0;

var _extends2 = _interopRequireDefault(_extends._extends.exports);

var _propTypes = _interopRequireDefault(index.propTypes.exports);

var _constants = constants.__exports;

var _utils = index$1.__exports;

var _propTypes2 = propTypes$1.__exports;

var pickerPropTypes = (0, _extends2.default)({}, _propTypes2.animationPropTypes, {
  classPrefix: _propTypes.default.string,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  locale: _propTypes.default.object,
  appearance: _propTypes.default.oneOf(['default', 'subtle']),
  block: _propTypes.default.bool,
  containerPadding: _propTypes.default.number,
  container: _propTypes.default.any,
  disabled: _propTypes.default.bool,
  toggleComponentClass: _propTypes.default.elementType,
  menuClassName: _propTypes.default.string,
  menuStyle: _propTypes.default.object,
  placeholder: _propTypes.default.node,
  placement: _propTypes.default.oneOf(_constants.PLACEMENT),

  /**
   * Prevent floating element overflow
   */
  preventOverflow: _propTypes.default.bool,
  open: _propTypes.default.bool,
  defaultOpen: _propTypes.default.bool,
  cleanable: _propTypes.default.bool,
  renderExtraFooter: _propTypes.default.func,
  renderValue: _propTypes.default.func,
  positionRef: _utils.refType,
  onOpen: _propTypes.default.func,
  onClose: _propTypes.default.func,
  onClean: _propTypes.default.func,

  /** @deprecated Use `onClose` instead */
  onHide: _propTypes.default.func
});
propTypes.__exports.pickerPropTypes = pickerPropTypes;
var listPickerPropTypes = (0, _extends2.default)({}, pickerPropTypes, {
  data: _propTypes.default.array,
  valueKey: _propTypes.default.string,
  labelKey: _propTypes.default.string,
  childrenKey: _propTypes.default.string,
  disabledItemValues: _propTypes.default.array,
  value: _propTypes.default.any,
  defaultValue: _propTypes.default.any,
  onChange: _propTypes.default.func
});
propTypes.__exports.listPickerPropTypes = listPickerPropTypes;
var pickerDefaultProps = {
  cleanable: true,
  placement: 'bottomStart',
  appearance: 'default'
};
propTypes.__exports.pickerDefaultProps = pickerDefaultProps;
var listPickerDefaultProps = (0, _extends2.default)({}, pickerDefaultProps, {
  data: [],
  disabledItemValues: [],
  childrenKey: 'children',
  valueKey: 'value',
  labelKey: 'label'
});
propTypes.__exports.listPickerDefaultProps = listPickerDefaultProps;

exports.default = propTypes.__exports;
