// rsuite/lib/Ripple/Ripple.js
'use strict';

var Ripple = require('../../../../_virtual/Ripple.js_commonjs-module');
require('../../../@babel/runtime/helpers/interopRequireWildcard.js');
require('../../../@babel/runtime/helpers/interopRequireDefault.js');
require('../../../@babel/runtime/helpers/extends.js');
require('../../../@babel/runtime/helpers/objectWithoutPropertiesLoose.js');
require('../../../@babel/runtime/helpers/inheritsLoose.js');
var React = require('react');
require('../../../prop-types/index.js');
var cx = require('classnames');
require('../Animation/Transition.js');
require('../../../dom-lib/es/animation/cancelAnimationFramePolyfill.js');
require('../../../dom-lib/es/animation/nativeRequestAnimationFrame.js');
require('../../../dom-lib/es/animation/requestAnimationFramePolyfill.js');
require('../../../dom-lib/es/transition/translateDOMPositionXY.js');
require('../../../dom-lib/es/getVendorPrefixedName.js');
require('../../../dom-lib/es/utils/emptyFunction.js');
require('../../../dom-lib/es/normalizeWheel.js');
var index$1 = require('../../../../_virtual/index2.js_commonjs-proxy');
require('../utils/index.js');
var Transition = require('../../../../_virtual/Transition.js_commonjs-exports');
var index$2 = require('../../../../_virtual/index.js_commonjs-exports');
var _extends = require('../../../../_virtual/extends.js_commonjs-module');
var objectWithoutPropertiesLoose = require('../../../../_virtual/objectWithoutPropertiesLoose.js_commonjs-module');
var inheritsLoose = require('../../../../_virtual/inheritsLoose.js_commonjs-module');
var index = require('../../../../_virtual/index3.js_commonjs-module');
var interopRequireWildcard = require('../../../../_virtual/interopRequireWildcard.js_commonjs-module');
var interopRequireDefault = require('../../../../_virtual/interopRequireDefault.js_commonjs-module');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

(function (module, exports) {

var _interopRequireWildcard = interopRequireWildcard.interopRequireWildcard.exports;

var _interopRequireDefault = interopRequireDefault.interopRequireDefault.exports;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(_extends._extends.exports);

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(objectWithoutPropertiesLoose.objectWithoutPropertiesLoose.exports);

var _inheritsLoose2 = _interopRequireDefault(inheritsLoose.inheritsLoose.exports);

var React = _interopRequireWildcard(React__default['default']);

var _propTypes = _interopRequireDefault(index.propTypes.exports);

var _classnames = _interopRequireDefault(cx__default['default']);

var _Transition = _interopRequireDefault(Transition.__exports);

var _domLib = index$1['default'];

var _utils = index$2.__exports;

var Ripple =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(Ripple, _React$Component);

  function Ripple(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.triggerRef = void 0;
    _this.mousedownListener = null;

    _this.getPosition = function (event) {
      var offset = (0, _domLib.getOffset)(_this.triggerRef.current);
      var offsetX = (event.pageX || 0) - offset.left;
      var offsetY = (event.pageY || 0) - offset.top;
      var radiusX = Math.max(offset.width - offsetX, offsetX);
      var radiusY = Math.max(offset.height - offsetY, offsetY);
      var radius = Math.sqrt(Math.pow(radiusX, 2) + Math.pow(radiusY, 2));
      return {
        width: radius * 2,
        height: radius * 2,
        left: offsetX - radius,
        top: offsetY - radius
      };
    };

    _this.handleMouseDown = function (event) {
      var position = _this.getPosition(event);

      var onMouseDown = _this.props.onMouseDown;

      _this.setState({
        rippling: true,
        position: position
      });

      onMouseDown === null || onMouseDown === void 0 ? void 0 : onMouseDown(position, event);
    };

    _this.handleRippled = function () {
      _this.setState({
        rippling: false
      });
    };

    _this.addPrefix = function (name) {
      return (0, _utils.prefix)(_this.props.classPrefix)(name);
    };

    _this.state = {
      rippling: false,
      position: {}
    };
    _this.triggerRef = React.createRef();
    return _this;
  }

  var _proto = Ripple.prototype;

  _proto.componentDidMount = function componentDidMount() {
    if (this.triggerRef.current) {
      this.mousedownListener = (0, _domLib.on)(this.triggerRef.current.parentNode, 'mousedown', this.handleMouseDown);
    }
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.mousedownListener) {
      this.mousedownListener.off();
    }
  };

  _proto.render = function render() {
    var _this$props = this.props,
        className = _this$props.className,
        classPrefix = _this$props.classPrefix,
        rest = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["className", "classPrefix"]);
    var classes = (0, _classnames.default)(this.addPrefix('pond'), className);
    var _this$state = this.state,
        position = _this$state.position,
        rippling = _this$state.rippling;
    var unhandled = (0, _utils.getUnhandledProps)(Ripple, rest);
    return React.createElement("span", (0, _extends2.default)({}, unhandled, {
      className: classes,
      ref: this.triggerRef
    }), React.createElement(_Transition.default, {
      in: rippling,
      enteringClassName: this.addPrefix('rippling'),
      onEntered: this.handleRippled
    }, function (props, ref) {
      var className = props.className,
          transitionRest = (0, _objectWithoutPropertiesLoose2.default)(props, ["className"]);
      return React.createElement("span", (0, _extends2.default)({}, transitionRest, {
        ref: ref,
        className: (0, _classnames.default)(classPrefix, className),
        style: position
      }));
    }));
  };

  return Ripple;
}(React.Component);

Ripple.propTypes = {
  classPrefix: _propTypes.default.string,
  className: _propTypes.default.string,
  onMouseDown: _propTypes.default.func
};

var _default = (0, _utils.defaultProps)({
  classPrefix: 'ripple'
})(Ripple);

exports.default = _default;
module.exports = exports.default;
}(Ripple.Ripple, Ripple.Ripple.exports));
