// react-day-picker/build/LocaleUtils.js
import { __exports as LocaleUtils } from '../../../_virtual/LocaleUtils.js_commonjs-exports';

Object.defineProperty(LocaleUtils, "__esModule", {
  value: true
});
LocaleUtils.formatDay = formatDay;
LocaleUtils.formatMonthTitle = formatMonthTitle;
LocaleUtils.formatWeekdayShort = formatWeekdayShort;
LocaleUtils.formatWeekdayLong = formatWeekdayLong;
LocaleUtils.getFirstDayOfWeek = getFirstDayOfWeek;
LocaleUtils.getMonths = getMonths;
var WEEKDAYS_LONG = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

var WEEKDAYS_SHORT = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'];

var MONTHS = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

function formatDay(day) {
  return day.toDateString();
}

function formatMonthTitle(d) {
  return MONTHS[d.getMonth()] + ' ' + d.getFullYear();
}

function formatWeekdayShort(i) {
  return WEEKDAYS_SHORT[i];
}

function formatWeekdayLong(i) {
  return WEEKDAYS_LONG[i];
}

function getFirstDayOfWeek() {
  return 0;
}

function getMonths() {
  return MONTHS;
}

LocaleUtils.default = {
  formatDay: formatDay,
  formatMonthTitle: formatMonthTitle,
  formatWeekdayShort: formatWeekdayShort,
  formatWeekdayLong: formatWeekdayLong,
  getFirstDayOfWeek: getFirstDayOfWeek,
  getMonths: getMonths
};
